grug/app/src/gas/storage.rs

```diff
@@ -147,7 +149,99 @@ where
         min: Option<Bound<K>>,
         max: Option<Bound<K>>,
         order: Order,
-    ) -> StdResult<Box<dyn Iterator<Item = StdResult<(K::Output, T)>> + 'b>> {
+    ) -> StdResult<Box<dyn Iterator<Item = StdResult<(K::Output, T)>> + 'b>>
+    where
+        T: 'b,
+    {
+        // Gas cost for creating an iterator.
+        gas_tracker.consume(GAS_COSTS.db_scan, "db_scan")?;
+
+        let iter = self
+            .range_raw(storage, min, max, order)
+            .metered(gas_tracker)
+            .map(|record| {
+                let (k_raw, v_raw) = record?;
+                let k = K::from_slice(&k_raw)?;
+                let v = C::decode(&v_raw)?;
+                Ok((k, v))
+            });
+
+        Ok(Box::new(iter))
+    }
+
+    fn save_with_gas(
+        &self,
+        storage: &mut dyn Storage,
+        gas_tracker: GasTracker,
+        key: K,
+        value: &T,
+    ) -> StdResult<()> {
+        let data_raw = C::encode(value)?;
+        let path = self.path(key);
+
+        let gas_cost = GAS_COSTS
+            .db_write
+            .cost(data_raw.len() + path.as_path().storage_key().len());
+
+        // Charge gas before writing the data, such that if run out of gas,
+        // the data isn't written.
+        gas_tracker.consume(gas_cost, "db_write")?;
+
+        path.as_path().save_raw(storage, &data_raw);
+
+        Ok(())
+    }
```

Comment:
For `save_with_gas` of indexed map, it seems you're using the same algo as normal map? This doesn't account for writing to the index set.